### PR TITLE
Remove unneeded mhlo.scatter normalization failure

### LINKED ISSUE #11839: Unsupported mhlo.scatter case
### What happened?

During testing a unet_batch model, this scatter type is not converted into linalg.

There are many other scatters, so only this type of scatter seems unsupported.

Here is the input file: https://gist.github.com/okkwon/99a89ff71a6484dab786494770bf9bb5

### Steps to reproduce your issue

```
iree-compile --iree-input-type=mhlo scatter.mlir --iree-hal-target-backends=cuda -o out.vmfb
```

### LINKED ISSUE #11888: mhlo.scatter fails to lower in iree-mhlo-to-linalg-ext pass
### What happened?

`mhlo.scatter` conversion fails because it does not have canonical dimensions. I think this is the same problem as #9361.

My use case is exporting a Stable Diffusion model training from Jax. I isolated a minimal reproducer from the original MLIR file.

### Steps to reproduce your issue

```
iree-compile \
    illegal_stablehlo_scatter.mlir \
    --iree-input-type=mhlo \
    --iree-hal-target-backends=cuda \
    -o illegal_stablehlo_scatter.vmfb
```

`illegal_stablehlo_scatter.mlir`:
```
func.func @main(%arg0: tensor<1x64x32x640xf32>, %arg1: tensor<64x1xi32>, %arg2: tensor<1x64x64x640xf32>) -> tensor<1x64x32x640xf32> {
  %1 = "stablehlo.scatter"(%arg0, %arg1, %arg2) ({
  ^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):
    %2 = stablehlo.add %arg3, %arg4 : tensor<f32>
    stablehlo.return %2 : tensor<f32>
  }) {indices_are_sorted = false, scatter_dimension_numbers = #stablehlo.scatter<update_window_dims = [0, 1, 3], inserted_window_dims = [2], scatter_dims_to_operand_dims = [2], index_vector_dim = 1>, unique_indices = false} : (tensor<1x64x32x640xf32>, tensor<64x1xi32>, tensor<1x64x64x640xf32>) -> tensor<1x64x32x640xf32>
  return %1 : tensor<1x64x32x640xf32>
}
```

This results in the error
```
illegal_stablehlo_scatter.mlir:12:8: error: failed to legalize operation 'mhlo.scatter' that was explicitly marked illegal
  %1 = "stablehlo.scatter"(%arg0, %arg1, %arg2) ({
       ^
illegal_stablehlo_scatter.mlir:12:8: note: see current operation: 
%0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ({
^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):
  %1 = "mhlo.add"(%arg3, %arg4) : (tensor<f32>, tensor<f32>) -> tensor<f32>
  "mhlo.return"(%1) : (tensor<f32>) -> ()
}) {indices_are_sorted = false, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0, 1, 3], inserted_window_dims = [2], scatter_dims_to_operand_dims = [2], index_vector_dim = 1>, unique_indices = false} : (tensor<1x64x32x640xf32>, tensor<64x1xi32>, tensor<1x64x64x640xf32>) -> tensor<1x64x32x640xf32>
// -----// IR Dump After ConvertMHLOToLinalgOnTensors Failed (iree-mhlo-to-linalg-on-tensors) //----- //
module {
  func.func @main(%arg0: tensor<1x64x32x640xf32>, %arg1: tensor<64x1xi32>, %arg2: tensor<1x64x64x640xf32>) -> tensor<1x64x32x640xf32> {
    %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ({
    ^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):
      %1 = mhlo.add %arg3, %arg4 : tensor<f32>
      mhlo.return %1 : tensor<f32>
    }) {indices_are_sorted = false, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0, 1, 3], inserted_window_dims = [2], scatter_dims_to_operand_dims = [2], index_vector_dim = 1>, unique_indices = false} : (tensor<1x64x32x640xf32>, tensor<64x1xi32>, tensor<1x64x64x640xf32>) -> tensor<1x64x32x640xf32>
    return %0 : tensor<1x64x32x640xf32>
  }
}
```

### What component(s) does this issue relate to?

_No response_

### Version information

This error appears on top of commit 343084fda533c48104aec0cf87e5ebacd1725122.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rsuderman**
A failure was included during mhlo.scatter's normalization that assumed the 0-th update dimension should never be updated This was incorrect and could be removed. Included a test to verify this behavior.

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 1]
```diff
@@ -511,11 +511,11 @@ struct ScatterOpImplicitBatch : public OpRewritePattern<mhlo::ScatterOp> {
...
-    auto indicesTy = indices.getType().cast<ShapedType>();
+    auto indicesTy = indices.getType().dyn_cast<RankedTensorType>();
 
     // Check whether indices has no batch dimension.
-    if (!indicesTy.hasRank()) return failure();
-    if (indicesTy.getRank() != 1 && indexVectorDim != 0) {
+    if (!indicesTy) return failure();
+    if (indicesTy.getRank() != 1 || indexVectorDim != 0) {
       return rewriter.notifyMatchFailure(op,
                                          "No implicit batch dimension to add.");
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:520]: lower case start while here.

File: compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_scatter.mlir [hunk 2]
```diff
@@ -209,3 +209,66 @@ func.func @scatter_operand_map(%arg0: tensor<5x4x1xi32>, %arg1: tensor<1x2xi32>,
...
+// CHECK: %[[EXPANDED:.+]] = tensor.expand_shape %[[ARG2]]
+// CHECK-SAME{literal}: [[0], [1], [2, 3], [4]] : tensor<44x1x64x640xf32> into tensor<44x1x64x1x640xf32>
+// CHECK: %[[SCATTER:.+]] = "mhlo.scatter"(%[[ARG0]], %[[TRANSPOSE]], %[[EXPANDED]])
+// CHECK: ^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):
+// CHECK:   %2 = mhlo.add %arg3, %arg4 : tensor<f32>
+// CHECK:   mhlo.return %2 : tensor<f32>
+// CHECK: indices_are_sorted = false
+// CHECK-SAME: scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [1, 2, 3, 4], scatter_dims_to_operand_dims = [2], index_vector_dim = 1>
+// CHECK-SAME: unique_indices = false
+// CHECK: return %[[SCATTER]] : tensor<1x64x32x640xf32>
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_scatter.mlir:274]: Newline?

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 3]
```diff
@@ -694,13 +789,13 @@ struct ScatterMaterializeInsertedDim
     auto operand = op.getInputs().front();
     auto indicesTy = indices.getType().cast<ShapedType>();
     auto operandTy = operand.getType().cast<ShapedType>();
+    auto dimNumbers = op.getScatterDimensionNumbers();
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:792]: Why is this change needed?

**Review by jpienaar — APPROVED** [review]
Nice, just small nits.

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: Done.

File: compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_scatter.mlir (hunk 2 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_scatter.mlir]: Done.

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 3 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: It is not required so I reverted this part of the change.

**Pull request merged**